Add GPU.sb3 extension

derpygamer2142 opened the pull request:
This adds blocks to use compute shaders in a Turbowarp project. This uses WebGPU(which is experimental!), and only supports compute shaders. From what little research I've done it seems like it would be a pain to add any of the other types, so I'm not doing that.

- Documentation [here](https://extensions.derpygamer2142.com/docs/gpusb3)
- Demos [here](https://extensions.derpygamer2142.com/demos)
- WGSL spec [here](https://www.w3.org/TR/WGSL/)
- WebGPU spec [here](https://www.w3.org/TR/webgpu/)
- WebGPU implementation status by browser [here](https://webgpu.io/)
- WebGPU fundamentals [here](https://webgpufundamentals.org/)

Notes:
- The only thing that's somewhat undocumented(I think, let me know if I forgot something) is buffer alignment and stuff, but webgpufundamentals does a better job of that.
- This doesn't work in the packager with electron. Not sure why, a quick google search leads me to [believe it's in beta](https://github.com/electron/electron/issues/26944).

![gpusb3](https://github.com/user-attachments/assets/2fc9812b-d8f3-4b7d-8a3c-f3c475f404a1)

derpygamer2142 committed:
add gpu.sb3

derpygamer2142 committed:
fix shaderNotFound bug

derpygamer2142 committed:
add sample

derpygamer2142 committed:
reformat, remove sample because ig i don't know how it works

Thebloxers998 commented:
It fails to get WebGPU adapter

Thebloxers998 commented:
What does the extension do

derpygamer2142 commented:
> It fails to get WebGPU adapter

what browser are you on, and what os

derpygamer2142 commented:
> What does the extension do

it's like pen+ but with numbers instead of colors and it all happens at the same time

Thebloxers998 commented:
> > It fails to get WebGPU adapter
> 
> what browser are you on, and what os

Chrome, Android, Why did you ask

Thebloxers998 commented:
> > What does the extension do
> 
> it's like pen+ but with numbers instead of colors and it all happens at the same time

That would be more hard but ok

derpygamer2142 commented:
> > > It fails to get WebGPU adapter
> > 
> > 
> > what browser are you on, and what os
> 
> Chrome, Android, Why did you ask?

WebGPU is experimental and doesn't work on some devices. Seems like android wouldn't be supported based on webgpu.io.

derpygamer2142 committed:
dual license

Thebloxers998 commented:
> > > > It fails to get WebGPU adapter
> > > 
> > > 
> > > what browser are you on, and what os
> > 
> > 
> > Chrome, Android, Why did you ask?
> 
> WebGPU is experimental and doesn't work on some devices. Seems like android wouldn't be supported based on webgpu.io.

Oh ok

FurryR commented:
well this is rough, make it more readable.

LilyMakesThings commented on images/derpygamer2142/gpusb3.svg:
no.

derpygamer2142 replied:
man

derpygamer2142 replied:
is it too silly or is it the intellectual property infringment

derpygamer2142 replied:
is the new thumbnail acceptable?

LilyMakesThings replied:
yes

hablethedev replied:
yay my epic thumbnail :D

LilyMakesThings commented on extensions/derpygamer2142/gpusb3.js:
```diff
@@ -0,0 +1,4488 @@
+break;
+};
+
+`); // GOD FUCKING DAMN IT "break if (condition)" IS IN THE OFFICIAL SPEC WHY THE HELL IS IT INVALID THIS IS BULLSHIT
```
Language.

derpygamer2142 replied:
😔😔😔😔😔😔

derpygamer2142 commented:
> well this is rough, make it more readable.

that's fair

derpygamer2142 committed:
"""fix""" thumbnail

FurryR commented:
I'm excited to see the extension coming out as it means that we could now perform GPGPU calculations on TurboWarp, meaning that 3D applications will be blazingly fast and AI is now possible.

derpygamer2142 committed:
readability and new block

derpygamer2142 commented:
added a bunch of comments, removed profanity, and I think I reordered stuff in a way that makes more sense but I don't remember

derpygamer2142 commented:
i have finally updated documentation to add the 3 new blocks

derpygamer2142 committed:
cool new colorz

derpygamer2142 committed:
formatting 🤫🧏

derpygamer2142 committed:
error handling fixes

derpygamer2142 committed:
fix operator_add, async issues

derpygamer2142 committed:
missing device error handling

important: this commit makes pretty much all the webgpu blocks async, so there's a 1 frame delay when they are run or something like that

derpygamer2142 commented:
The latest commit(https://github.com/TurboWarp/extensions/pull/1762/commits/84c7b87ca44ab8815907336960d6518894e34ac2) adds error handling in case the device was lost or the project was run before it was finished getting the adapter+device. this makes pretty much every webgpu block async, which is pretty cringe. I could alternatively:
 - use .thens on the promise instead of awaiting it, removing the need for async(? not sure if this would work)
 - check if the device is invalid, and if it is just don't run the block while also throwing an error
 - do it anyways and add a way to check if the last block run threw an error

derpygamer2142 committed:
change how invalid devices are handled

derpygamer2142 commented:
chose the last option as suggested
https://discord.com/channels/837024174865776680/1132437506655268926/1337565380243099679

derpygamer2142 committed:
fix cringe missing error handling

derpygamer2142 committed:
fix array buffer colors

derpygamer2142 committed:
fromatahdswtv7gug

derpygamer2142 committed:
fix some colors and stuff

derpygamer2142 committed:
translations and formatting

derpygamer2142 commented:
i need to make some changes to the gpu.sb3 compiler so it doesn't get broken by translations

derpygamer2142 committed:
bug fixes

derpygamer2142 committed:
rework compiler

derpygamer2142 committed:
formatting skill issue?

derpygamer2142 committed:
third times the charm

derpygamer2142 commented:
changes complete, this is ready for review

yuri-kiss commented on extensions/derpygamer2142/gpusb3.js:
```diff
@@ -0,0 +1,4875 @@
+// Name: GPU.sb3
+// ID: gpusb3
```
this id should contain a form of your name

yuri-kiss commented on extensions/derpygamer2142/gpusb3.js:
```diff
@@ -0,0 +1,4875 @@
+
+  let buffersExt;
+  // load exposed extension stuff
+  Scratch.vm.runtime.on("EXTENSION_ADDED", () => {
```
maybe use runtime.off when you finally get the right extension

yuri-kiss commented on extensions/derpygamer2142/gpusb3.js:
```diff
@@ -0,0 +1,4875 @@
+  Scratch.vm.runtime.on("EXTENSION_ADDED", () => {
+    buffersExt = Scratch.vm.runtime.ext_0znzwBuffers; // miyo's buffer extension exposes some stuff, you can use these in the arraybuffer blocks for convenience
+  });
+  let shaders = {};
+  let error = {};
+  let resources = {
+    buffers: {},
+    bindGroups: {},
+    bindGroupLayouts: {},
+    bufferRefs: {}, // deprecated, not used anywhere anymore
+    arrayBuffers: {},
+    views: {},
+    textures: {}, // webgpu texture objects, actual images will be yoinked from ~~the pen+ costume library(if available)~~(scrapped idea, too complicated. i go into more detail elsewhere) and costume list
+    samplers: {}, // this doesn't work with compute shaders but i kept it in case i lock in and add other stuff(extremely unlikely but deleting code causes me pain)
+  };
```
these should probably be constants

derpygamer2142 replied:
it makes it easier to reset and is used to reduce duplicate code in the bind group entry block:
```js
    bindGroupEntry(args, util) {
      ...
      const kv = {
        buffer: "buffers",
        storageTexture: "textures",
        sampler: "samplers",
      }; // bind group entry type -> resources key

      // the object to bind to that slot. buffers are freaky and need a special object
      let o;
      const type = kv[args.TYPE] ?? "buffers";
      if (
        !Object.prototype.hasOwnProperty.call(
          resources[type],
          Scratch.Cast.toString(args.RESOURCE)
        )
      ) {
        return this.throwError(
          "ResourceNotFound",
          "The specified resource doesn't exist",
          "BindGroupEntryBlock",
          `Either the resource type is invalid or the provided resource name doesn't exist.`,
          util
        );
      }
      if (type == "buffers") {
        o = {
          buffer: buffers[args.RESOURCE],
        };
        //console.log(o)
      } else {
        o = resources[type][args.RESOURCE];
        if (type == "textures") o = o.createView();
      }
      
      ...
      }

```

yuri-kiss replied:
its less clear and can be prone to bugs

yuri-kiss commented on extensions/derpygamer2142/gpusb3.js:
```diff
@@ -0,0 +1,4875 @@
+      // @ts-ignore
+      if (!navigator.gpu) {
+        // why angry red lines >: (
+        return this.throwError(
+          "WebGPUUnavailable",
+          "WebGPU is not supported",
+          "Init",
+          "WebGPU is unavailable in the current context",
+          util
+        );
```
you should find a better way to tell the user that webGPU isnt available

derpygamer2142 replied:
this was requested by a couple people as opposed to showing an alert and not working anymore so the project can gracefully fall back on something else

yuri-kiss replied:
you should still alert the user if the project isnt packaged

yuri-kiss commented on extensions/derpygamer2142/gpusb3.js:
```diff
@@ -0,0 +1,4875 @@
+     * @param {import("scratch-vm").BlockUtility} util util
+     */
+    throwError(errorname, errorbody, errorsource, full, util) {
+      error = {
+        name: errorname ?? "Undefined. This is an error, please report it!",
+        body: errorbody ?? "Undefined. This is an error, please report it!",
+        source: errorsource ?? "Undefined. This is an error, please report it!",
+        full: full ?? "Undefined. This is an error, please report it!",
+      };
```
just add these props on the error object instead of reassigning it, its clearer if you do

derpygamer2142 replied:
like
```js
throwError(name, body, source, full, util) {
      name = name ?? "Undefined. This is an error, please report it!";
      body = body ?? "Undefined. This is an error, please report it!";
      source = source ?? "Undefined. This is an error, please report it!";
      full = full ?? "Undefined. This is an error, please report it!";
      
      error = { name, body, source, full };
      
      ...
}
```

yuri-kiss replied:
no?

derpygamer2142 replied:
k then i'm not really sure what you mean

yuri-kiss commented on extensions/derpygamer2142/gpusb3.js:
```diff
@@ -0,0 +1,4875 @@
+      };
+
+      console.error(error);
+      if (util) {
+        util.startHats("gpusb3_onError");
+      } else {
+        Scratch.vm.runtime.startHats("gpusb3_onError");
+      }
```
these functions dont differ, there is no need to use util here

yuri-kiss commented on extensions/derpygamer2142/gpusb3.js:
```diff
@@ -0,0 +1,4875 @@
+        color1: "#27c90e",
+        // color2: "#166af2",
+        // color3: "#032966",
+        docsURI: "https://extensions.derpygamer2142.com/docs/gpusb3",
```
make this the would be tw docs page instead please

yuri-kiss reviewed (changes requested):
yeah this is going to be a multipart review if i get around to it, this is giant and out of the scope of my knowledge as I dont know webgpu

derpygamer2142 committed:
make requested changes

derpygamer2142 committed:
alert when device has issues

derpygamer2142 committed:
replace resources object with individual things

derpygamer2142 commented:
i'm planning on adding some utilities for struct building on the cpu and maybe possible perhaps some rendering stuff, let me know if you have any ideas as to what this should look like

derpygamer2142 commented:
vadik came in clutch and told me that I can just put the texture of a skin to a canvas that was rendered to via webgpu, I'm going to add that

Brackets-Coder commented:
Is this still a thing or is it stagnant?

derpygamer2142 commented:
> Is this still a thing or is it stagnant?

it is, i'm just waiting for this pr to be reviewed and i'm working on adding rendering

Brackets-Coder commented:
> > > > Is this still a thing or is it stagnant?
> > > 
> > > 
> > > it is, i'm just waiting for this pr to be reviewed and i'm working on adding rendering
> > 
> > 
> > Okay this one is bigger so I'll review it when I have time
> 
> thats what they all said

Okay, I'll tell you what. I'll get around to it at my earliest convenience, likely sometime today

Brackets-Coder commented on extensions/derpygamer2142/gpusb3.js:
```diff
@@ -0,0 +1,4794 @@
+// ID: derpygamergpusb3
+// Description: Use WebGPU compute shaders to accelerate your projects.
+// By: derpygamer2142 <https://scratch.mit.edu/users/insanetaco2000/>
+// License: MIT
```
Are you using MIT out of personal preference or because you're using a library that uses it?

Brackets-Coder replied:
Okay, since we don't have strict requirements I suppose this is acceptable.

Brackets-Coder commented on extensions/derpygamer2142/gpusb3.js:
```diff
@@ -0,0 +1,4794 @@
+  let buffers = {},
+    bindGroups = {},
+    bindGroupLayouts = {},
+    bufferRefs = {}, // deprecated, not used anywhere anymore
```
If it's deprecated why didn't you remove it

derpygamer2142 replied:
done

Brackets-Coder commented on extensions/derpygamer2142/gpusb3.js:
```diff
@@ -0,0 +1,4794 @@
+    bufferRefs = {}, // deprecated, not used anywhere anymore
+    arrayBuffers = {},
+    views = {},
+    textures = {}, // webgpu texture objects, actual images will be yoinked from ~~the pen+ costume library(if available)~~(scrapped idea, too complicated. i go into more detail elsewhere) and costume list
+    samplers = {}; // this doesn't work with compute shaders but i kept it in case i lock in and add other stuff(extremely unlikely but deleting code causes me pain)
```
Okay, these seem unfinished but since it causes you emotional damage I suppose you can leave them

derpygamer2142 replied:
it's planned now anyways

Brackets-Coder commented on extensions/derpygamer2142/gpusb3.js:
```diff
@@ -0,0 +1,4794 @@
+      samplers = {}; // also unused, clear just in case
+
+      // @ts-ignore
+      if (!navigator.gpu) {
+        // why angry red lines >: (
+        if (typeof scaffolding === "undefined")
+          alert(
+            "WebGPU is unavailable in the current context. Check https://webgpu.io for more information."
+          );
+        return this.throwError(
+          "WebGPUUnavailable",
+          "WebGPU is not supported",
+          "Init",
+          "WebGPU is unavailable in the current context"
+        );
+        // throw new Error("WebGPU is not supported.");
+      }
```
Nice check, does the website educate users on whether their browser has experimental support as feature flags? A lot of people say Safari doesn't support WebGPU but I'm extremely familiar with macOS and it's actually hidden behind a "Stable" feature flag after enabling developer settings...

Brackets-Coder commented on extensions/derpygamer2142/gpusb3.js:
```diff
@@ -0,0 +1,4794 @@
+        this.device = null; // requestDevice will never return null so we need to account for that ourselves
+      });
+
+      // note to self: uncomment this on release
+      this.device.addEventListener("uncapturederror", (event) => {
+        this.throwError(
+          "UnclassifiedError",
+          // @ts-ignore
+          event.error.message,
+          "Unknown",
+          // @ts-ignore
+          event.error
+        ); // this is literally in the spec and the mdn docs, idk why it's complaining about event.error being undefined https://www.w3.org/TR/webgpu/#eventdef-gpudevice-uncapturederror
+      });
+    }
```
This is your reminder to uncomment

derpygamer2142 replied:
it's already uncommented so it's fine

Brackets-Coder commented on extensions/derpygamer2142/gpusb3.js:
```diff
@@ -0,0 +1,4794 @@
+     *
+     * @param {"buffers" | "bindGroups" | "bindGroupLayouts" | "arrayBuffers" | "views" | "textures" | "samplers"} type
+     */
+    resourceFromType(type) {
+      switch (type) {
+        case "buffers":
+          return buffers;
+        case "bindGroups":
+          return bindGroups;
+        case "bindGroupLayouts":
+          return bindGroupLayouts;
+        case "arrayBuffers":
+          return arrayBuffers;
+        case "views":
+          return views;
+        case "textures":
+          return textures;
+        case "samplers":
+          return samplers; // unused, returned anyways
+        default:
+          return null;
+      }
+    }
```
Unless I'm wrong can you make them part of an object and do something like this

```suggestion
    resourceFromType(type) {
        return resources[type];
    }
```
If it's not possible that's okay but it might help clean it up a lot since all of the strings are exactly the same as the name of the returned values.

derpygamer2142 replied:
that was how it was originally but a reviewer asked for it to be changed

Brackets-Coder commented on extensions/derpygamer2142/gpusb3.js:
```diff
@@ -0,0 +1,4794 @@
+        name: Scratch.translate("GPU.sb3"),
+
+        color1: "#27c90e",
+        // color2: "#166af2",
+        // color3: "#032966",
```
Were these important?

derpygamer2142 replied:
no, i originally got rid of it so the colors were exactly the same in the docs as the actual extension. i'll remove those comments

Brackets-Coder commented on extensions/derpygamer2142/gpusb3.js:
```diff
@@ -0,0 +1,4794 @@
+            ),
+            isEdgeActivated: false,
+            arguments: {
+              // all arguments here are grabbed using some workspace tomfoolery, hence why they don't support anything other than text
+              // originally these were fetched using a modified version of the wgsl transpiler(generateWGSL) but i just had stupid idiot brain fungus
+              // and thought that was a good idea despite it being horrible and the least efficient way to do it
```
Clarify what you mean?

derpygamer2142 replied:
idk how this vscode review extension works or if it waits to post my response until after i've commited but i am making a change here

Brackets-Coder commented on extensions/derpygamer2142/gpusb3.js:
```diff
@@ -0,0 +1,4794 @@
+            ),
+            arguments: {
+              GPUFUNC: {
+                // GPUFUNC is an old name that i am to scared to change
```
if GPUFUNC is a bad name it doesn't hurt to change

Brackets-Coder commented on extensions/derpygamer2142/gpusb3.js:
```diff
@@ -0,0 +1,4794 @@
+              GPUFUNC: {
+                // GPUFUNC is an old name that i am to scared to change
+                type: Scratch.ArgumentType.STRING,
+                defaultValue: Scratch.translate("myShader"),
+              },
+              BINDGROUP: {
+                type: Scratch.ArgumentType.STRING,
+                defaultValue: Scratch.translate("myBindGroup"),
```
@CubesterYT Were we supposed to be translating the default values because I never have

CubesterYT replied:
Usually yeah

In some cases, no

Brackets-Coder commented on extensions/derpygamer2142/gpusb3.js:
```diff
@@ -0,0 +1,4794 @@
+          },
+
+          {
+            hideFromPalette: true,
```
Why is this block hidden?

derpygamer2142 replied:
samplers are not available in compute shaders

Brackets-Coder commented on extensions/derpygamer2142/gpusb3.js:
```diff
@@ -0,0 +1,4794 @@
+          "ReadBufferBlock",
+          "Failed to run block as the connection the GPU is invalid."
+        );
+      // WARNING:
+      // MAY CONTAIN BAD IDEA JUICE
+      // GPUMapMode.READ assumes no writing will be done
+      // aka if you want to write to your mapped buffer you need to transfer it to the cpu, mess with it, then transfer it to a different buffer
+      // and send it back to the gpu
```
That sounds... complicated and unperformant, again, probably should be in the documentation

Brackets-Coder commented on extensions/derpygamer2142/gpusb3.js:
```diff
@@ -0,0 +1,4794 @@
+      }[format];
+    }
+
+    textureType() {
+      return "This block allows you to add texture types to your shaders";
+    }
+
+    samplerType() {
+      return "This block allows you to add sampler types to your shaders";
+    }
```
Do these blocks actually do anything?

Brackets-Coder commented on extensions/derpygamer2142/gpusb3.js:
```diff
@@ -0,0 +1,4794 @@
+      });
+    }
+
+    declareStruct() {
+      return;
+    }
+    structProperty() {
+      return;
+    }
+    structType() {
+      return "This block lets you use a struct as a type.";
+    }
```
Do these blocks actually do anything?

Brackets-Coder commented on images/derpygamer2142/gpusb3.png:
That looks an awful lot like an Nvidia GPU, and you can't really have text in your banner due to translation and typography...

derpygamer2142 replied:
it is a brandless and definitely not traced gpu and i think the .sb3 is fine because it's a file extension, but i'll ask if the "by" part can be removed

Brackets-Coder commented on package-lock.json:
```diff
@@ -12,6 +12,7 @@
         "@turbowarp/json": "^0.1.2",
         "@turbowarp/scratchblocks": "^3.6.5",
         "@turbowarp/types": "git+https://github.com/TurboWarp/types-tw.git#tw",
+        "@webgpu/types": "^0.1.51",
```
is webgpu types absolutely required for this extension

derpygamer2142 replied:
it is borderline impossible to develop this extension without it but it is not required for the functionality

Brackets-Coder reviewed (commented):
This extension is so large and beyond my pay grade I don't think I'll be of much help.
This review is by far the biggest one I've done and I'm just skimming the surface.
As I've said in one of these comments, I'm more familiar with rendering libraries than the lowest-level of the APIs...

Unfortunately I haven't made time to test this yet but it seems like this one will be quite an endeavor.

Brackets-Coder commented:
I'd recommend putting your documentation at extensions.turbowarp.com, just because some users may not trust your personal site.

Also, considering how much this extension appears to be a work-in-progress, I'm going to convert it to a draft, **but if I'm wrong please feel free to open it up again.**

derpygamer2142 commented on extensions/derpygamer2142/gpusb3.js:
```diff
@@ -0,0 +1,4794 @@
+                type: Scratch.ArgumentType.STRING,
+                menu: "FUNCTYPES",
+                defaultValue: "void",
+              } /*,
+                            ARGS: {
+                                type: Scratch.ArgumentType.STRING,
+                                defaultValue: ""
+                            }*/,
+            },
```
i did that to keep people from typing in the argument slot, i could probably change that but i'm still not sure how much to "childproof" this extension.

derpygamer2142 commented on extensions/derpygamer2142/gpusb3.js:
```diff
@@ -0,0 +1,4794 @@
+              "textureDimensions",
+              "textureLoad",
+              "textureStore",
+              // only for fragment shaders
+              //"textureSample",
+              // "textureSampleBaseClampToEdge"
+              // "textureSampleBias"
+              // "textureSampleCompare"
+              // "textureSampleCompareLevel"
```
no, that is planned though

derpygamer2142 commented on extensions/derpygamer2142/gpusb3.js:
```diff
@@ -0,0 +1,4794 @@
+              "Int32Array",
+              "Uint32Array",
+              "Float32Array",
+              // webgpu doesn't support types below this comment, but people may want to access them using those types anyways
+              "Int8Array",
+              "Uint8Array",
+              "Uint8ClampedArray",
+              "Int16Array",
+              "Uint16Array",
+              // "Float16Array",
+              // float16array is only available in firefox so it is excluded for compatibility
+              "BigInt64Array",
+              "BigUint64Array",
+              "Float64Array",
```
that was already requested by a now former reviewer and i was planning on doing it after all the documentation was done

derpygamer2142 commented on extensions/derpygamer2142/gpusb3.js:
```diff
@@ -0,0 +1,4794 @@
+     * @returns
+     */
+    textFromOp(util, id, unsafe) {
+      // IMPORTANT: All menus in this extension accept reporters because otherwise the field gets stored elsewhere
+      // and it's a whole thing.
```
pretty sure i already cast what's needed, and accepting reporters is another part of the question about child proofing

derpygamer2142 commented on extensions/derpygamer2142/gpusb3.js:
```diff
@@ -0,0 +1,4794 @@
+                case 'e ^': return Math.exp(n);
+                case '10 ^': return Math.pow(10, n);
+                */
+            let op = "How do you mess up this badly?"; // this shouldn't ever be present unless you intentionally do it
```
maps the operator_mathop block to various math operations, the comment is for reference of the difference options

derpygamer2142 commented on extensions/derpygamer2142/gpusb3.js:
```diff
@@ -0,0 +1,4794 @@
+              );
+              return "Unexpected input";
+            }
+            code = code.concat(this.genWGSL(util, b.NAME, recursionDepth + 1)); // if your variable name is invalid it's your own fault and you can cry about it
```
anything besides a-Z, numbers and underscores or having the variable name start with a number

derpygamer2142 commented on extensions/derpygamer2142/gpusb3.js:
```diff
@@ -0,0 +1,4794 @@
+
+    c_arbitraryWGSL() {}
+
+    r_arbitraryWGSL() {
+      return "This block allows you to add custom WGSL to your shaders.";
+    }
+
+    atomicType() {
+      return "This block returns a type that can be used to declare an atomic variable.";
+    }
+
+    atomicLoad() {
+      return "This block gets an atomic function";
+    }
+
+    c_atomicFunc() {}
+
+    r_atomicFunc() {
+      return "This block performs a thread safe operation on an atomic variable.";
+    }
+
+    barrier() {}
+
+    variablePointer() {
+      return "This block converts a variable to a pointer. Equivilant to *someVar in c.";
+    }
```
reporters return a short blurb of documentation for some usability and the functions existing prevents errors

derpygamer2142 commented on extensions/derpygamer2142/gpusb3.js:
```diff
@@ -0,0 +1,4794 @@
+      let j;
+      try {
+        j = JSON.parse(args.ARRAY);
+        if (!Array.isArray(j)) throw new Error("skibidi toilet ohio rizz");
```
oh this is the only one that's here unintentionally i will fix that

derpygamer2142 commented on extensions/derpygamer2142/gpusb3.js:
```diff
@@ -0,0 +1,4794 @@
+          return blob.fields.TEXTUREBASETYPES.value;
+        }
+
+        case "gpusb3_samplerType": {
+          // v past me either didn't finish this comment or removed part of it, and i don't remember what it said
+          // i am now reworking the compiler and i could probably fix whatever i was yapping about but i don't want to look through stuff to figure it out
+          // the "solution" here is to just do whatever this case does and return the desired input if anyone ever decides to use this kind of block again
+
+          // IMPORTANT: For blocks with no inputs, such as the scrapped sampler type block,
+          return "sampler"; // the codesmell here is crazy but this should work well enough
+        }
```
i remembered what i was writing there, fixed

derpygamer2142 commented on extensions/derpygamer2142/gpusb3.js:
```diff
@@ -0,0 +1,4794 @@
+      try {
+        j = JSON.parse(args.ARRAY);
+        if (!Array.isArray(j))
+          throw new Error("balkan rage winter arc jonkler trollge phonk"); // 🤫🧏
```
fixed

derpygamer2142 committed:
make requested changes

derpygamer2142 committed:
forgot to run format

derpygamer2142 commented on extensions/derpygamer2142/gpusb3.js:
```diff
@@ -0,0 +1,4794 @@
+              ">>=",
+              "<<=",
+            ],
+          },
```
done

derpygamer2142 commented on extensions/derpygamer2142/gpusb3.js:
```diff
@@ -0,0 +1,4794 @@
+        }
+        case "gpusb3_menu_RAWTYPES": {
+          return blob.fields.RAWTYPES.value;
+        }
+
+        case "gpusb3_menu_BGLENTRYTYPES": {
+          return blob.fields.BGLENTRYTYPES.value;
+        }
+
+        case "gpusb3_menu_CONSTRUCTABLETYPES": {
+          return blob.fields.CONSTRUCTABLETYPES.value;
+        }
+
+        case "gpusb3_menu_BUFFERUSAGE": {
+          return blob.fields.BUFFERUSAGE.value;
+        }
+
+        case "gpusb3_menu_BUFFERENTRYTYPE": {
+          return blob.fields.BUFFERENTRYTYPE.value;
+        }
+
+        case "gpusb3_menu_VARUSAGE": {
+          return blob.fields.VARUSAGE.value;
+        }
+
+        case "gpusb3_menu_ATOMICBASES": {
+          return blob.fields.ATOMICBASES.value;
+        }
+
+        case "gpusb3_menu_ATOMICFUNCTIONS": {
+          return blob.fields.ATOMICFUNCTIONS.value;
+        }
+
+        case "gpusb3_menu_BARRIERFUNCTIONS": {
+          return blob.fields.BARRIERFUNCTIONS.value;
+        }
+
+        case "gpusb3_menu_TEXTUREBASETYPES": {
+          return blob.fields.TEXTUREBASETYPES.value;
+        }
```
done

derpygamer2142 committed:
change textfromop function
